paddle/fluid/operators/filter_instag_op.h

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstring>
+#include <random>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include "paddle/fluid/framework/eigen.h"
+#include "paddle/fluid/framework/lod_tensor.h"
+#include "paddle/fluid/framework/mixed_vector.h"
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/memory/memcpy.h"
+#include "paddle/fluid/platform/assert.h"
+
+namespace paddle {
+namespace operators {
+using Tensor = framework::Tensor;
+using SelectedRows = framework::SelectedRows;
+using LoDTensor = framework::LoDTensor;
+#if defined(PADDLE_WITH_CUDA)
+template <typename T>
+using Vector = framework::Vector<T>;
+#else
+template <typename T>
+using Vector = framework::CPUVector<T>;
+#endif
+
+template <typename T>
+class FilterInstagKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    // X1 is global FC output
+    // Dim [batch size, embedding size]
+    auto* x1 = context.Input<LoDTensor>("X1");
+    // X2 is ins tag list
+    // LoD [[0, Sum(ins1), Sum(ins1, ins2), ... ]]
+    auto* x2 = context.Input<LoDTensor>("X2");
+    // X3 is local fc tag list
+    // LoD [[0, Sum(fc1), Sum(fc1, fc2) ...]]
+    auto* x3 = context.Input<Tensor>("X3");
+
+    std::unordered_set<int64_t> filter_tag;
+    auto* x3_data = x3->data<int64_t>();
+    size_t len = x3->dims()[0];
+    for (size_t i = 0; i < len; i++) {
+      filter_tag.insert(x3_data[i]);
+    }
+
+    // expected auto = const int64_t
+    auto* x2_data = x2->data<int64_t>();
+    // e.g get [0, 1, 2, 3, ...]
+    auto x2_lods = x2->lod()[0];
+    auto x1_lods = x1->lod()[0];
+
+    std::unordered_map<int64_t, int64_t> mmap_aux;
+    std::vector<size_t> ins_after_filter;
+    Vector<size_t> out_lods(1, 0);
+    for (size_t i = 0; i < x2_lods.size() - 1; i++) {
+      for (size_t j = x2_lods[i]; j < x2_lods[i + 1]; j++) {
+        if (filter_tag.find(x2_data[j]) != filter_tag.end()) {
+          ins_after_filter.push_back(x2_lods[i]);
+          size_t batch_len = x1_lods[i + 1] - x1_lods[i];
+          mmap_aux[out_lods.back()] = x1_lods[i];
+          out_lods.push_back(out_lods.back() + batch_len);
```

Comment:
what if filter_tag is empty after this for loop? do you just return a lodtensor with batch size of zero?

Reply:
yes, now it is empty. And loss weight will indicate it is empty.